New package: llvm18

Calandracas606 opened the pull request:
<!-- Uncomment relevant sections and delete options which are not applicable -->

#### Testing the changes
- I tested the changes in this PR: **briefly**


#### New package
- This new package conforms to the [package requirements](https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#package-requirements): **YES**


<!-- Note: If the build is likely to take more than 2 hours, please add ci skip tag as described in
https://github.com/void-linux/void-packages/blob/master/CONTRIBUTING.md#continuous-integration
and test at least one native build and, if supported, at least one cross build.
Ignore this section if this PR is not skipping CI.
-->

#### Local build testing
- I built this PR locally for my native architecture, x86_64-musl
- I built this PR locally for these architectures (if supported. mark crossbuilds):
  - x86_64-musl
  - i686
  - aarch64 (cross)
  - aarch64-musl (native + cross)
  - armv7l (cross)
  - armv7l-musl (cross)
  - armv6l (cross)
  - armv6l-musl (cross)

[ci skip]

sgn commented on srcpkgs/llvm18/template:
```diff
@@ -0,0 +1,725 @@
+		vmove "usr/lib/libclang*.a"
+		vmove "usr/lib/libclang*.so"
+		vmove usr/share/clang
+		vmove usr/bin/hmaptool
+		vmove usr/bin/git-clang-format
```
`hmaptool` goes to `-tools-extra`, `git-clang-format` goes to same packages with `clang-format`.

sgn replied:
Honestly, I think `hmaptool` should go to clang

sgn commented on srcpkgs/llvm18/template:
```diff
@@ -0,0 +1,725 @@
+build_wrksrc=llvm
+build_style=cmake
+_ext_suffix=".cpython-${py3_ver/./}-linux-${XBPS_TARGET_LIBC/glibc/gnu}.so"
+pycompile_dirs="usr/share/scan-view"
```
This `pycompile_dirs` goes to `clang-analyzer`

zlice commented:
I saw this while checking on where zig was at. I have [intel opencl](https://github.com/void-linux/void-packages/pull/43084) that was working with llvm15 but from intel's open issues 16, and I assume 17, will be in the air for a while. With 18 being updated, should I just mark intel opencl as a lost cause? I don't know that they will ever be "caught up" and things will be able to compile with void's versions.

Calandracas606 commented:
> I saw this while checking on where zig was at. I have [intel opencl](https://github.com/void-linux/void-packages/pull/43084) that was working with llvm15 but from intel's open issues 16, and I assume 17, will be in the air for a while. With 18 being updated, should I just mark intel opencl as a lost cause? I don't know that they will ever be "caught up" and things will be able to compile with void's versions.

llvm15 won't be going anywhere. the libraries (libllvm15, libclang15, etc) will likely stick around as long as they are needed, llvm12 is still packaged for example

zlice commented:
I'll try again with 15 sometime, I thought the `-devel` package was needed and that seems to be gone.

zlice commented:
So it may not be llvm/clang needed devel, but SPIRV devel packages. I assume that conflicts between versions which is why only the latest (17) is in the repos?

leahneukirchen commented:
Please add libomp-devel to the dependency hull of llvm18-devel, this breaks cmake often.

Calandracas606 commented:
> Please add libomp-devel to the dependency hull of llvm18-devel, this breaks cmake often.

that fix should already be part of this PR

Calandracas606 commented:
template seems mostly fine.

only remaining issue is that compiler-rt17 fails to build on `armv*-musl` due to failing a static assertion. (works fine with compiler-rt18)

seems to only occur when the only runtime built is compiler-rt. if libcxx and unwinder are built alongside it, the error does not occur.

Calandracas606 commented:
> So it may not be llvm/clang needed devel, but SPIRV devel packages. I assume that conflicts between versions which is why only the latest (17) is in the repos?

@zlice at one point I did have a working intel level-zero and intel opencl template with llvm15, but never got around to figuring out how to patch it for musl

it probably makes sense to version SPIRV, there's no reason why libLLVMSPRIVLib.so.18 and libLLVMSPIRVLib.so.17 can't co-exist.

Calandracas606 commented:
I think the PR is nearly ready to go, just doing some final build testing

Calandracas606 commented:
PR is ready for merge.

Should be relatively safe, it only adds llvm18, but doesn't actually build anything against it. Having it available ahead of time will be useful for when rust and chromium are eventually dependent on it

Calandracas606 commented:
rebased the branch.

A review would be really appreciated.

I've been running this PR for months now, and I'm pretty sure I've fixed all the quirks, and that this should be safe to merge

moeishome commented:
agree it would be appreciated to have this reviewed!

chrysos349 commented:
thanks for updating llvm to v18!

why does the shared lib `libclang18` depend on include headers `clang18-headers`? i think it should not. or am i wrong?
p.s. the same applies to the current 17 version.

Calandracas606 commented:
pushed a small fix for when natively compiling on aarch64
